# HG changeset patch
# User kamg
# Date 1336075036 14400
# Node ID f64381365f7cdc3cde6d2db3b850e39109ee9764
# Parent  17aa937e8207323719df4dac99cea86da5d4833f
7160757: Problem with hotspot/runtime_classfile
Summary: Allow only current and super invokespecials of <init>
Reviewed-by: never, coleenp, dcubed

diff --git a/src/share/vm/classfile/verifier.cpp b/src/share/vm/classfile/verifier.cpp
--- openjdk/hotspot/src/share/vm/classfile/verifier.cpp
+++ openjdk/hotspot/src/share/vm/classfile/verifier.cpp
@@ -1871,10 +1871,10 @@
   VerificationType type = current_frame->pop_stack(
     VerificationType::reference_check(), CHECK_VERIFY(this));
   if (type == VerificationType::uninitialized_this_type()) {
-    // The method must be an <init> method of either this class, or one of its
-    // superclasses
+    // The method must be an <init> method of this class or its superclass
+    klassOop superk = current_class()->super();
     if (ref_class_type.name() != current_class()->name() &&
-        !name_in_supers(ref_class_type.name(), current_class())) {
+        ref_class_type.name() != superk->klass_part()->name()) {
       verify_error(bci, "Bad <init> method call");
       return;
     }
